### ISSUE TITLE
How to define a one way converter?

### ISSUE THREAD

**Issue by szguoxz** [linq2db/linq2db#4368]
i.e., converters only take effect from .net to DB,
and converters only take effect from DB to .Net.
Currently, the converters work both ways.

**Comment by igor-tkachev** [comment]
Witch way?

**Comment by szguoxz** [comment]
Both ways.
I want to define a set of converters which only take effect from Dotnet to sql, and another set of converters only take effect from sql to Dotnet

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by igor-tkachev** [comment]
> Both ways. I want to define a set of converters which only take effect from Dotnet to sql, and another set of converters only take effect from sql to Dotnet 

Can you provide an example of what you need? Like "I have XXX type in DB and want to have it as YYY in data model". And other way around.

**Comment by szguoxz** [comment]
I have a date time data type in db. Historic reason, there is no time zone info in db. I always convert my date time to utc and save it to db.

My developers sometimes will create a date time type in Dotnet without a time zone, which is unspecified date time kind. He actually means datetimekind.local.

So I have this requirement:
When saving to db, convert all datetime type with kind of unspecified to local datetime, then to utc and save to db.
When load from db, all datetime Is unspecified since time zone information was Not saved. I have to convert all datetime Unspecified to utc then to local.

Basically how we should treat unspecified datetime.
From Dotnet to sql, treat unspecified as datetimekind.local
From db to Dotnet, treat unspecified As datetimekind.utc

Not sure if you can follow me here. :-)

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by igor-tkachev** [comment]
Check the following example:

```c#
using System;
using System.Diagnostics;

using LinqToDB;
using LinqToDB.Data;
using LinqToDB.Mapping;

namespace L2DBDateTimeTest
{
	internal class Program
	{
		static DataOptions _options = new DataOptions()
			.UseSQLiteMicrosoft("Data Source=db.sqlite")
			.UseTracing(TraceLevel.Info, info => Console.WriteLine(info.SqlText));

		// Use specific MappingSchema for DateTime conversion or set MappingSchema.Default.
		//
		static MappingSchema _utcMappingSchema = new ("UtcConvert");

		static Program()
		{
			// Converts DateTime to UTC when reading from DB.
			//
			_utcMappingSchema.SetConvertExpression<DateTime,DateTime>     (dt => dt.ToLocalTime());
			// Converts DateTime to UTC when writing to DB.
			//
			_utcMappingSchema.SetConvertExpression<DateTime,DataParameter>(dt => new DataParameter() { Value = dt.ToUniversalTime(), DataType = DataType.DateTime });
		}

		static void Main(string[] args)
		{
			{
				using var db = new DataConnection(_options);

				db.AddMappingSchema(_utcMappingSchema);

				db.DropTable<Table>(tableOptions: TableOptions.CheckExistence);

				var t = db.CreateTable<Table>(tableOptions: TableOptions.CheckExistence);

				// Test different scenarios.
				//

				db.Insert(new Table { Date = DateTime.Now });

				t
					.Value(t => t.Date, DateTime.Now)
					.Insert();

				t.BulkCopy([new Table { Date = DateTime.Now }]);

				db.InlineParameters = true;

				db.Insert(new Table { Date = DateTime.Now });

				t
					.Value(t => t.Date, DateTime.Now)
					.Insert();

				t.BulkCopy([new Table { Date = DateTime.Now }]);
			}

			{
				// Check that DateTime is converted to UTC when reading from DB.
				// Do not use 'UtcConvert' MappingSchema.
				//
				using var db = new DataConnection(_options);

				foreach (var r in db.GetTable<Table>())
				{
					Console.WriteLine(r.Date);
				}
			}
		}

		public class Table
		{
			[Column, DataType(DataType.DateTime)]
			public DateTime Date { get; set; }
		}
	}
}
```

**Comment by szguoxz** [comment]
I tested it out in my usecase, it won't work.  Apparently, it won't work anyway.

first of all:

SetConvertExpression, to be honestm I don't know if does anything different than SetConverter. I don't know the difference anyway.

And I tried it, it works both ways. So the result is not right.

Calvin
________________________________

**Comment by szguoxz** [comment]
The first converter works both ways

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by igor-tkachev** [comment]
Can I see your example showing that it is not working?

**Comment by szguoxz** [comment]
I am not in office, will get it to you later. Thanks!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by szguoxz** [comment]
> Can I see your example showing that it is not working?

I looked closely to your example, the converters are different. Here is my simplified use case:

var ctx = new AppDataContext();
var a = new Authtoken { Appid = 100, AuthtokenColumn = Guid.NewGuid().ToString(), Expirationtime = DateTime.SpecifyKind(DateTime.Now, DateTimeKind.Unspecified) };
Console.WriteLine(a.ToJson());
ctx.Insert(a);
var b = ctx.Get<Authtoken>(x => x.Appid == 100 && x.AuthtokenColumn == a.AuthtokenColumn);
Console.WriteLine(b.Expirationtime.Value.Kind);
Console.WriteLine(b.Expirationtime);

My goal is to make sure a.ExpirationTime  == b.ExpirationTime.
----------------------------
The difficulty here is still the same, thee converters works both ways. I have stepped through your example, The first converter got called from Dotnet to SQL, and from SQL to Dotnet. The second converter got called only from DotNetToSQL, that's because we do not use DataParameter type in normal dotnet code.

Are you trying to solve my problem while still using the two-way converter? I have already solve it using a different approach going around a long way and with some hack into  dotnet. But really it's not a good solution. The ideal way is for LinqToDB to support one way converter. 
Your second converter use datatype match effectively created a one way converter from Dotnet To SQL. But there seems no  easy way to create a SQL to DOTNET one way converter.
